src/main/java/org/opensearch/knn/indices/ModelIndex.java

```diff
@@ -0,0 +1,231 @@
+                KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+        );
+
+        IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+        indexRequestBuilder.setId(modelId);
+        indexRequestBuilder.setSource(parameters);
+
+        put(indexRequestBuilder, listener);
+    }
+
+    /**
+     * Put a model into the system index. Non-blocking. When no id is passed in, OpenSearch will generate the id
+     * automatically. The id can be retrieved in the IndexResponse.
+     *
+     * @param modelBlob byte array of model
+     * @param listener handles index response
+     */
+    public void put(KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
+        String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+
+        Map<String, Object> parameters = ImmutableMap.of(
+                KNNConstants.KNN_ENGINE, knnEngine.getName(),
+                KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+        );
+
+        IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+        indexRequestBuilder.setSource(parameters);
+
+        put(indexRequestBuilder, listener);
+    }
+
+    private void put(IndexRequestBuilder indexRequestBuilder, ActionListener<IndexResponse> listener) {
+        if (!isCreated()) {
+            throw new IllegalStateException("Cannot put model in index before index has been initialized");
+        }
+
+        // Fail if the id already exists. Models are not updateable
+        indexRequestBuilder.setOpType(DocWriteRequest.OpType.CREATE);
+        indexRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        indexRequestBuilder.execute(listener);
+    }
+
+    /**
+     * Get a model from the system index. Call blocks.
+     *
+     * @param modelId to retrieve
+     * @return byte array representing the model
+     * @throws ExecutionException thrown on search
+     * @throws InterruptedException thrown on search
+     */
+    public byte[] get(String modelId) throws ExecutionException, InterruptedException {
+        if (!isCreated()) {
+            throw new IllegalStateException("Cannot get model \"" + modelId + "\". Model index does not exist.");
+        }
+
+        /*
+            GET /<model_index>/<modelId>?source_includes=<model_blob>&_local
+        */
+        GetRequestBuilder getRequestBuilder = new GetRequestBuilder(client, GetAction.INSTANCE, MODEL_INDEX_NAME)
+                .setId(modelId)
+                .setFetchSource(KNNConstants.MODEL_BLOB_PARAMETER, null)
+                .setPreference("_local");
+        GetResponse getResponse = getRequestBuilder.execute().get();
+
+        Object blob = getResponse.getSourceAsMap().get(KNNConstants.MODEL_BLOB_PARAMETER);
+
+        if (blob == null) {
+            throw new IllegalArgumentException("ModelID: \"" + modelId + "\" is not present in index");
```

Comment:
Should this message be more user focused? Something like ```There is no model available with the provided Id?```

Reply:
Sure, I can update.